### DISCUSSION TITLE
[Feature]: Greenhouse block

### DISCUSSION THREAD

**Discussion by Manyaarvadiya** [MCreator-Examples/Tale-of-Biomes#85] [Feature Requests]
### Feature information

need to make a oxygen distributer block for my mod so please bro

**Comment by northwesttreesgaming** [comment]
I had plans to make a greenhouse for crops, though it's for plants it could be adapted for an oxygen system. However, I will need to figure this out when the time comes because at the moment I am not sure how to do it and I need to finish up the rice crop mechanics before I can start working on this which has been on hold until other systems are in place.

But I will work towards this in the future for sure.
I have marked this as to do so I can keep track of the progress.

**Comment by Manyaarvadiya** [comment]
Any ideas or logic on how this could work. As this is the only thing left in my mod. And mod jam deadline is near and still can't figure it out

There is code for it in ad astra GitHub but I don't know how to read Java or any stuff so like please

**Comment by northwesttreesgaming** [comment]
I have not brainstormed any ideas on how to do this yet nor have I done it so I can't really say what would work or not.
With that said I have a few tutorials grass spreading mechanics and even other blocks that replace things over time.
This could be used to spread invisible blocks that when the or entity is in them give them NBT for what ever oxygen level you have.

You would also want to make an air-tight block and set a rough distance which any six sides of the location the air would spread to is in reach of an air-tight block, this way natural shelters like caves would not provide a place for it to spread to as this could cause major lag in post generation conditions.

The last thing you would need is some kind of source I guess the source would generate air blocks.
As far as leaking mechanics I have no idea how this would work. Maybe you could test if there were gaps but this goes into coding. I am no programmer so I have no clue how this would be done but as an airtight room, this would at least give you the foundations for it. 3D objects may also require custom versions depending on how the player can walk in them just keep this in mind as in some cases non-air pockets could cause the air blocks to go through. You may want to try making them skip 1 block, in this case, this way it has a chance to jump obstructions but other than that I have no idea how to do it.

Greenhouse it would be less important to make a sealed room all I would need to do is test if the blocks are next to it and the plants are indoors. This could technically be applied the same way but this would in some cases court as outdoors if the player or entity was on the other side of the blocks with one block above. Just something to keep in mind I guess.

In cases like this, I wish we had a way to make custom air blocks like cave air and such but we don't at the moment.

**Comment by Manyaarvadiya** [comment]
Well we can make air blocks. Just set it's texture to a transparent texture which you can make by erasing an existing texture. Then set it's bounding box to 0 and make it replaceable which is a tick box. And wallah it's a air block

**Comment by northwesttreesgaming** [comment]
Yea I see your point, though 3D models still render the spaces as air, doing it this way would not fill those spaces with your block you would need custom blocks for that or do it through conditions to make it count which is what I was trying to say.

Water acts much the same way as air in that regard as it works with 3D models, you can still breathe in them even when standing next to them if you place a fence in your gas it should be able to keep the gas while still providing the 3D model. Sadly though placing a fence would just add a regular fence with actual air.

So a bit different mechanics than actual air or water.

**Comment by Manyaarvadiya** [comment]
So long till when. Did you think about this. I am still waiting to finish my mod due to this

**Comment by northwesttreesgaming** [comment]
I only record once every two weeks, not sure when I can finish the current project but I will see if I can fit it in after that.
It may be ideal for me to just make an example workspace.

**Comment by northwesttreesgaming** [comment]
I got a project workspace done today, spent 4 hours working on it.
I just want to be vary clear that I don't normally help people in this manner for their mods.
It takes a lot of time to do, but giving we both had pretty much the same thing planned I had nothing to lose for putting a  workspace together. I made it public [over here](https://github.com/MCreator-Examples/Projects/discussions/146).

With that being said I did add notes to each procedure, however, the procedure that controls the effect for the player assumes you have some kind of oxygen or NBT variable system for that you can adapt the ApplyEffect procedure inside the repeat block that controls the SizeZ reading.

It's also possible to set block NBT if you need to update blocks to read as in an oxygen room.
This would work much the same way you would just carry over the local nX, nY, and nZ math blocks and replace the X, Y, and Z blocks for your NBT variables which I will most likely end up doing for my greenhouse effect.

Anyhow I hope this helps and is not too hard for you to learn how to set up. Most of it is pretty advanced but make sure you check the global variables and tags tabs as there are some important settings in them as well as the block and GUI I implemented too.

If you get stuck then you will have to wait until I can do a tutorial which will be based on what the community wants to fast-track for a tutorial. I do advanced tutorials every second week so no clue when it would end up being recorded.

**Reply by Manyaarvadiya** [reply]
it all works great but there is no automated detection. i want so it is place and run so you dont need to input values

**Reply by Manyaarvadiya** [reply]
> it all works great but there is no automated detection. i want so it is place and run so you dont need to input values

K but did you read this msg

**Comment by Manyaarvadiya** [comment]
Thx a lot very very lot man. Also just a question. How do I switch the dimention of player along with it's player and keep player ridding the vehicle

**Reply by northwesttreesgaming** [reply]
I honestly don't know if making them ride the entity again once in the other dimension is possible but you can swap entities including the entity ridding to a new dimension.

There should be a selector block for the entity being ridden in the entity data tab.
But when it comes to making them ride again, I think the UUID of the entity is changed in the other world so it may not be possible to get them automatically ridding the entity again, also it takes less time for an entity to load than the player due to the rendering so its quite possible that the entity will wander a bit before the player has a chance to load in fully.

**Reply by Manyaarvadiya** [reply]
My ship doesn't wander. Ik I should not advertise this here but since you are helping me with this final bit I want you to see my creation by joining my discord. It doesn't have any members yet so you'll be the first one. https://discord.com/invite/26FW3DfT

Look under the welcome tab. Not to boast  but my mod is quite revolutionary

**Reply by northwesttreesgaming** [reply]
What I have said is what I know sadly don't know of any other work around, sorry.
Your best bet would be to ask on the MCreator forums.

**Reply by Manyaarvadiya** [reply]
K but did you read the msg replied to your main msg

**Reply by northwesttreesgaming** [reply]
Yes. :)
Your mod sound really good.

**Reply by Manyaarvadiya** [reply]
Thx :) but I mean look way above. It's above these replies.

Ill just type it here. The example of oxygen distributer works really well but I don't want it to ask for size and stuff I want to make it place and run . Meaning automatic so no input of values

**Reply by northwesttreesgaming** [reply]
Sorry, I have lots on my mind right now. I did not catch that message.
You could skip the right-click action and instead set the OffsetX, OffsetY, OffsetZ, SizeX, SizeY and SizeZ block NBT through the "When block added" block trigger, this will make it so when the block is added to the world it will apply static values automatically. However, not having a way to set it could limit the position of the block and the size of the area.

**Reply by Manyaarvadiya** [reply]
K sure but give me the default values that would work for atleast a room with size 1000 .when I say 1000 I mean that many blocks not actual size.
